File: scripts/publish_upsert.js
const fs = require('fs');
const path = require('path');
const { createClient } = require('@libsql/client');

function loadEnv() {
  const envPath = path.join(__dirname, '..', '.env.local');
  if (fs.existsSync(envPath)) {
    const envContent = fs.readFileSync(envPath, 'utf8');
    envContent.split('\n').forEach(line => {
      const trimmed = line.trim();
      if (!trimmed || trimmed.startsWith('#')) return;
      const parts = trimmed.split('=');
      if (parts.length >= 2) {
        const key = parts[0]?.trim();
        const value = parts.slice(1).join('=').trim().replace(/(^["']|["']$)/g, '');
        if (key) process.env[key] = value;
      }
    });
  }
}

loadEnv();

function slugify(text) {
  return text
    .toString()
    .toLowerCase()
    .trim()
    .replace(/\s+/g, '-')
    .replace(/[^\w\-]+/g, '')
    .replace(/\-\-+/g, '-')
    .replace(/^-+/, '')
    .replace(/-+$/, '');
}

function calculateReadingTime(htmlContent) {
  const wordsPerMinute = 225;
  const plainText = htmlContent.replace(/<[^>]*>/g, '');
  const words = plainText.trim().split(/\s+/).length;
  return Math.max(1, Math.ceil(words / wordsPerMinute));
}

async function publish() {
  let logs = [];
  const log = (...args) => logs.push(args.join(' '));

  const draftPath = path.join(__dirname, '..', 'draft_article.json');
  if (!fs.existsSync(draftPath)) {
    log('Error: Draft file not found');
    fs.writeFileSync(path.join(__dirname, '..', 'publish_upsert_log.txt'), logs.join('\n'), 'utf8');
    process.exit(1);
  }

  const raw = fs.readFileSync(draftPath, 'utf8');
  const draft = JSON.parse(raw);

  const dbUrl = process.env.TURSO_CONNECTION_URL;
  const dbToken = process.env.TURSO_AUTH_TOKEN;

  if (!dbUrl) {
    log('Error: TURSO_CONNECTION_URL is missing in .env.local');
    fs.writeFileSync(path.join(__dirname, '..', 'publish_upsert_log.txt'), logs.join('\n'), 'utf8');
    process.exit(1);
  }

  const client = createClient({ url: dbUrl, authToken: dbToken });

  const slug = slugify(draft.title);
  const readingTime = calculateReadingTime(draft.content);
  const now = Date.now();

  try {
    log(`[*] Checking if article slug "${slug}" exists in Turso database...`);
    const checkRes = await client.execute({
      sql: 'SELECT id, title FROM articles WHERE slug = ?',
      args: [slug]
    });

    if (checkRes.rows.length > 0) {
      const existingId = checkRes.rows[0].id;
      log(`[+] Found existing article with slug "${slug}" (ID: ${existingId}). Updating row...`);

      const updateQuery = `
        UPDATE articles SET
          title = ?,
          excerpt = ?,
          content = ?,
          category = ?,
          image = ?,
          author = ?,
          updated_at = ?,
          meta_title = ?,
          meta_description = ?,
          keywords = ?,
          reading_time = ?,
          og_image = ?,
          twitter_image = ?,
          published_by = ?
        WHERE slug = ?
      `;

      const updateArgs = [
        draft.title.trim(),
        draft.excerpt.trim(),
        draft.content.trim(),
        draft.category,
        draft.image.trim(),
        draft.author.trim(),
        now,
        draft.title.trim(),
        draft.excerpt.trim(),
        draft.keywords || draft.category.toLowerCase(),
        readingTime,
        draft.og_image || draft.image.trim(),
        draft.twitter_image || draft.image.trim(),
        draft.author.trim(),
        slug
      ];

      await client.execute({ sql: updateQuery, args: updateArgs });
      log(`[+] SUCCESS: Updated live article "${draft.title}" on database!`);
    } else {
      log(`[*] Article slug "${slug}" not found in database. Inserting new article...`);
      const newId = Math.random().toString(36).substring(2, 15);

      const insertQuery = `
        INSERT INTO articles (
          id, title, slug, excerpt, content, category, image, author,
          published_at, created_at, updated_at, featured, status,
          meta_title, meta_description, keywords, reading_time, views,
          tags, content_type, viral_score, source_name, source_url,
          views_7d, views_30d, og_image, twitter_image, published_by,
          research_ref, metadata
        ) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
      `;

      const insertArgs = [
        newId,
        draft.title.trim(),
        slug,
        draft.excerpt.trim(),
        draft.content.trim(),
        draft.category,
        draft.image.trim(),
        draft.author.trim(),
        now,
        now,
        now,
        draft.featured ? 1 : 0,
        'published',
        draft.title.trim(),
        draft.excerpt.trim(),
        draft.keywords || draft.category.toLowerCase(),
        readingTime,
        0,
        draft.tags || '',
        draft.content_type || 'news',
        draft.viral_score || 0,
        draft.source_name || '',
        draft.source_url || '',
        0,
        0,
        draft.og_image || draft.image.trim(),
        draft.twitter_image || draft.image.trim(),
        draft.author.trim(),
        draft.research_ref || '',
        draft.metadata || ''
      ];

      await client.execute({ sql: insertQuery, args: insertArgs });
      log(`[+] SUCCESS: Published new article "${draft.title}" to database!`);
    }

    // Ping Netlify revalidate endpoint if API key present
    const baseUrl = process.env.NEXT_PUBLIC_SITE_URL || 'https://kampusfilter.com';
    const revalidateWebhookUrl = `${baseUrl.replace(/\/$/, '')}/api/revalidate`;
    const notifySecret = process.env.KAMPUSFILTER_API_KEY;

    if (notifySecret) {
      try {
        await fetch(revalidateWebhookUrl, {
          method: 'POST',
          headers: {
            'Content-Type': 'application/json',
            'x-api-key': notifySecret,
          },
          body: JSON.stringify({ slug, category: draft.category }),
        });
        log('[+] Revalidation webhook triggered successfully.');
      } catch (e) {
        log('[!] Revalidation trigger note:', e.message);
      }
    }

  } catch (err) {
    log('[!] Database operation failed:', err.message);
  } finally {
    client.close();
    const outDir = "C:/Users/Admin/.gemini/antigravity/brain/fce01f4b-9b57-4b52-9044-1a0f0ab08850/scratch";
    if (!fs.existsSync(outDir)) fs.mkdirSync(outDir, { recursive: true });
    fs.writeFileSync(path.join(outDir, 'publish_result.txt'), logs.join('\n'), 'utf8');
    fs.writeFileSync(path.join(__dirname, '..', 'publish_upsert_log.txt'), logs.join('\n'), 'utf8');
  }
}

publish();

File: scripts/publish_upsert.py
#!/usr/bin/env python3
import os
import json
import urllib.request
import urllib.parse
import re
import time
import math

def load_env():
    env_path = os.path.join(os.path.dirname(__file__), '..', '.env.local')
    if os.path.exists(env_path):
        with open(env_path, 'r', encoding='utf-8') as f:
            for line in f:
                line = line.strip()
                if not line or line.startswith('#'):
                    continue
                parts = line.split('=', 1)
                if len(parts) == 2:
                    key = parts[0].strip()
                    val = parts[1].strip().strip('"').strip("'")
                    os.environ[key] = val

load_env()

def slugify(text):
    text = text.lower().strip()
    text = re.sub(r'\s+', '-', text)
    text = re.sub(r'[^\w\-]+', '', text)
    text = re.sub(r'\-\-+', '-', text)
    return text.strip('-')

def calculate_reading_time(html_content):
    plain_text = re.sub(r'<[^>]*>', '', html_content)
    words = len(plain_text.strip().split())
    return max(1, math.ceil(words / 225))

def main():
    log = []
    def log_print(msg):
        print(msg)
        log.append(msg)

    draft_path = os.path.join(os.path.dirname(__file__), '..', 'draft_article.json')
    if not os.path.exists(draft_path):
        log_print("[!] Error: draft_article.json does not exist.")
        return

    with open(draft_path, 'r', encoding='utf-8') as f:
        draft = json.load(f)

    db_url = os.environ.get('TURSO_CONNECTION_URL', '')
    db_token = os.environ.get('TURSO_AUTH_TOKEN', '')

    if not db_url or not db_token:
        log_print("[!] Error: TURSO_CONNECTION_URL or TURSO_AUTH_TOKEN is missing in .env.local")
        return

    http_url = db_url.replace('libsql://', 'https://')
    pipeline_url = f"{http_url}/v2/pipeline"
    headers = {
        "Authorization": f"Bearer {db_token}",
        "Content-Type": "application/json"
    }

    title = draft["title"].strip()
    slug = slugify(title)
    excerpt = draft["excerpt"].strip()
    content = draft["content"].strip()
    category = draft["category"].strip()
    image = draft["image"].strip()
    author = draft["author"].strip()
    reading_time = calculate_reading_time(content)
    now_ms = int(time.time() * 1000)

    log_print(f"[*] Processing article: '{title}'")
    log_print(f"[*] Target slug: '{slug}'")

    # Step 1: Check if article with slug exists
    check_payload = {
        "requests": [
            {
                "type": "execute",
                "stmt": {
                    "sql": "SELECT id, title FROM articles WHERE slug = ?",
                    "args": [{"type": "text", "value": slug}]
                }
            },
            {"type": "close"}
        ]
    }

    req = urllib.request.Request(
        pipeline_url,
        data=json.dumps(check_payload).encode('utf-8'),
        headers=headers,
        method='POST'
    )

    try:
        with urllib.request.urlopen(req) as resp:
            res_json = json.loads(resp.read().decode('utf-8'))
            rows = res_json['results'][0]['response']['result']['rows']

            if len(rows) > 0:
                article_id = rows[0][0]['value']
                log_print(f"[+] Found existing article (ID: {article_id}). Executing UPDATE query...")

                update_sql = """
                UPDATE articles SET
                    title = ?,
                    excerpt = ?,
                    content = ?,
                    category = ?,
                    image = ?,
                    author = ?,
                    updated_at = ?,
                    meta_title = ?,
                    meta_description = ?,
                    keywords = ?,
                    reading_time = ?,
                    og_image = ?,
                    twitter_image = ?,
                    published_by = ?
                WHERE slug = ?
                """
                update_args = [
                    {"type": "text", "value": title},
                    {"type": "text", "value": excerpt},
                    {"type": "text", "value": content},
                    {"type": "text", "value": category},
                    {"type": "text", "value": image},
                    {"type": "text", "value": author},
                    {"type": "integer", "value": str(now_ms)},
                    {"type": "text", "value": title},
                    {"type": "text", "value": excerpt},
                    {"type": "text", "value": draft.get("keywords", category.lower())},
                    {"type": "integer", "value": str(reading_time)},
                    {"type": "text", "value": draft.get("og_image", image)},
                    {"type": "text", "value": draft.get("twitter_image", image)},
                    {"type": "text", "value": author},
                    {"type": "text", "value": slug}
                ]

                update_payload = {
                    "requests": [
                        {
                            "type": "execute",
                            "stmt": {
                                "sql": update_sql,
                                "args": update_args
                            }
                        },
                        {"type": "close"}
                    ]
                }

                up_req = urllib.request.Request(
                    pipeline_url,
                    data=json.dumps(update_payload).encode('utf-8'),
                    headers=headers,
                    method='POST'
                )
                with urllib.request.urlopen(up_req) as up_resp:
                    log_print(f"[+] SUCCESS: Live article '{title}' successfully updated on Turso DB!")
            else:
                log_print(f"[*] Article slug '{slug}' not found. Executing INSERT query...")
                new_id = f"art_{now_ms}"

                insert_sql = """
                INSERT INTO articles (
                    id, title, slug, excerpt, content, category, image, author,
                    published_at, created_at, updated_at, featured, status,
                    meta_title, meta_description, keywords, reading_time, views,
                    tags, content_type, viral_score, source_name, source_url,
                    views_7d, views_30d, og_image, twitter_image, published_by,
                    research_ref, metadata
                ) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
                """

                insert_args = [
                    {"type": "text", "value": new_id},
                    {"type": "text", "value": title},
                    {"type": "text", "value": slug},
                    {"type": "text", "value": excerpt},
                    {"type": "text", "value": content},
                    {"type": "text", "value": category},
                    {"type": "text", "value": image},
                    {"type": "text", "value": author},
                    {"type": "integer", "value": str(now_ms)},
                    {"type": "integer", "value": str(now_ms)},
                    {"type": "integer", "value": str(now_ms)},
                    {"type": "integer", "value": "1" if draft.get("featured") else "0"},
                    {"type": "text", "value": "published"},
                    {"type": "text", "value": title},
                    {"type": "text", "value": excerpt},
                    {"type": "text", "value": draft.get("keywords", category.lower())},
                    {"type": "integer", "value": str(reading_time)},
                    {"type": "integer", "value": "0"},
                    {"type": "text", "value": draft.get("tags", "")},
                    {"type": "text", "value": draft.get("content_type", "news")},
                    {"type": "integer", "value": str(draft.get("viral_score", 0))},
                    {"type": "text", "value": draft.get("source_name", "")},
                    {"type": "text", "value": draft.get("source_url", "")},
                    {"type": "integer", "value": "0"},
                    {"type": "integer", "value": "0"},
                    {"type": "text", "value": draft.get("og_image", image)},
                    {"type": "text", "value": draft.get("twitter_image", image)},
                    {"type": "text", "value": author},
                    {"type": "text", "value": draft.get("research_ref", "")},
                    {"type": "text", "value": draft.get("metadata", "")}
                ]

                insert_payload = {
                    "requests": [
                        {
                            "type": "execute",
                            "stmt": {
                                "sql": insert_sql,
                                "args": insert_args
                            }
                        },
                        {"type": "close"}
                    ]
                }

                in_req = urllib.request.Request(
                    pipeline_url,
                    data=json.dumps(insert_payload).encode('utf-8'),
                    headers=headers,
                    method='POST'
                )
                with urllib.request.urlopen(in_req) as in_resp:
                    log_print(f"[+] SUCCESS: Live article '{title}' successfully published to Turso DB!")

    except Exception as e:
        log_print(f"[!] Database execution error: {e}")

    # Write log file
    log_out = os.path.join(os.path.dirname(__file__), '..', 'publish_result.txt')
    with open(log_out, 'w', encoding='utf-8') as f:
        f.write("\n".join(log))

if __name__ == "__main__":
    main()
